rfcs/ui-building-blocks.md

```diff
@@ -0,0 +1,413 @@
+
+We need to settle on a common standard, in order to enable interop between and establish basic building blocks that more complex UI decisions can build off of.
+
+## Guide-level explanation
+
+User interfaces in Bevy are made out of **widgets**, which are modified by the **styles** that are applied to them, and the aesthetic and functional behavior of is ultimately implemented through **UI systems**.
+
+A **widget** is a distinct part of the UI that needs its own data: be that its position, local state or any style parameters.
+Widgets in Bevy are represented by entities with a `Widget` marker component.
+
+Each widget has an associated `Styles` component, which contains a `Vec<Entity>` which points to some (possibly 0) number of **styles**, represented as entities with a `Style` marker component.
+
+Each style in that vector is applied, in order, to the widget, overriding the **style parameters** that already exist.
+Style parameters are stored as components, and serve to control some element of its presentation, such as the font, background color or text alignment.
+Style parameters can be reused across disparate widget types, with functionality achieved through systems that query for the relevant components.
+The type of the style parameter components determines which behavior it controls,
+while the value returned by `.get()` controls the final behavior of the widget.
+
+Every style parameter has both a `MyStyle<Base>` and a `MyStyle<Final>` variant, stored together on each entity.
+When creating a new style parameter, you must ensure that it implements `StyleParam`, typically achieved with `#[derive(StyleParam)]`.
+
+When working on complex games or applications, you're likely to want to group your styles into **themes**,
+automatically applying them to large groups of widgets at once.
+In Bevy, themes are applied by adding a generic system that corresponds to that theme to your app's schedule.
+Select a marker component type `W`, then add a theme system to your app with `W` as a type parameter:
+`app.add_startup_system(solarized_theme::<Button>.system())` will then create an entity that stores the solarized theme
+style parameters to your world, and adds the appropriate `Entity` reference to the end of each of the `Styles` components
+on all entities with `Widget` that have the `Button` marker component on them.
+
+Generally, you'll want to add these systems to a startup stage, but you may also find it useful to add them to various `State`s.
+This can work well for quickly toggling between various themes, like you might see in a light-dark mode.
+
+### Example: Building a Simple Widget
+
+```rust
+commands.spawn_bundle(ButtonBundle::default());
+```
+
+### Example: Building a Compound Widget
+
+```rust
+commands.spawn_bundle(ButtonBundle::default())
+  .with_children(|parent| {
+    parent.spawn_bundle(TextBundle::default())
```

Comment:
I would like to have a sugar over it, so instead of spawning bundles I as an user will write less code by invoking a member function `with_text`/`add_text` for `TextBundle`, for example.

```rust
trait SpawnText {
    fn spawn_text(&mut self, /* ... */) -> &mut Self
}

impl<'a, b'>  SpawnText for ChildBuilder<'a, 'b> {
    // ...
}
```
```rust
commands.spawn_bundle(ButtonBundle::default())
  .with_children(|parent| {
    parent.spawn_text("Hello, World!")
  });
```

Not sure that it's possible now due to lack of optional named parameter support.

Reply:
I agree with your dislike of the parent-child API. It's not great, but I copied the existing syntax to keep the scope of this RFC sane.

I'd like to avoid custom builder functions everywhere (like for `spawn_text`); they're a lot of boilerplate to have to write and maintain, and really obscure what's happening here in a way that makes it harder for people to develop an accurate mental model.

Reply:
So if I correctly understand you, the current API won't be touched by the proposal at all and what I'm asking is for another discussion, right?

Reply:
Correct. The parent-child API and the strategy for composing widgets is a separate discussion. IMO it should use relations (https://github.com/bevyengine/bevy/pull/1627), but there's a fair bit of design and then implementation work left to do there.